packages/loopring_v3/contracts/impl/libtransactions/WithdrawTransaction.sol

```diff
@@ -16,6 +16,13 @@ import "../libexchange/ExchangeWithdrawals.sol";
 
 /// @title WithdrawTransaction
 /// @author Brecht Devos - <brecht@loopring.org>
+/// @dev The following 4 types of withdrawals are supported:
+///      - withdrawType = 0: offchain withdrawals with EdDSA signatures
+///      - withdrawType = 1: offchain withdrawals with ECDSA signatures or onchain appprovals
+///      - withdrawType = 2: onchain valid forced withdrawals (owner and accountID matche)
+///      - withdrawType = 3: onchain invalid forced withdrawals (owner and accountID mismatch)
+/// TODO(brecht):
+///      - withdrawType = 4: offchain operator-initiated withdrawals for protocol fees or in shutdown mode.
```

Comment:
The different types are for the different behaviours in the circuit. `withdrawType = 4` like this would not behave differently than types 2/3 in the circuit (so we'd have to have 4/5, but they would do exactly the same thing).

The types are not related to how the onchain verfication is done. Type `1` withdrawals can be authorized onchain with an ECDSA signature or an approved hash, but the actual behaviour is exactly the same, it's just authorized differently, so they use the same type. The operator just needs to provide the correct authentication.

Reply:
I'lll keep it as-is